### PR TITLE
Add Gelato API Module

### PR DESCRIPTION
**Pull request by nlordell**
Follow up to #599 

I previously experimented with a Gelato relay service submitter in #599. The idea is to allow a separate relay service to execute settlement transactions on behalf of solvers so that they don't have to worry about execution details.

This PR is the first in a stack of 3 for enabling Gelato submission support. I am basically just breaking down the existing #599 into smaller pieces to make them easier to review and "self-contained".

In the following PRs I will:
1. Introduce the Solver trampoline contract and settlement signing.
2. Implement a Gelato submitter using the API introduced in this PR executed over the solver trampoline.

### Test Plan

Added a manual test for executing a transaction over the Gelato relay. Requires a 1Balance account with an API key.

```
% cargo test -p shared -- --ignored --nocapture gelato_api
   Compiling shared v0.1.0 (/Users/nlordell/Developer/cowprotocol/services/crates/shared)
    Finished test [unoptimized + debuginfo] target(s) in 14.19s
     Running unittests src/lib.rs (target/debug/deps/shared-b80690e295f197ca)

running 1 test
executing task 0xf75fd1215ac82c0b77b58bf0f51537184f621bba41db847cc7ba2a7920e4cca1
task is CheckPending...
task is CheckPending...
task is CheckPending...
task is WaitingForConfirmation...
task is WaitingForConfirmation...
task is WaitingForConfirmation...
task is WaitingForConfirmation...
task is WaitingForConfirmation...
task is WaitingForConfirmation...
task is WaitingForConfirmation...
task is WaitingForConfirmation...
task is WaitingForConfirmation...
task is WaitingForConfirmation...
task is WaitingForConfirmation...
task is WaitingForConfirmation...
task is WaitingForConfirmation...
task is WaitingForConfirmation...
task is WaitingForConfirmation...
task is WaitingForConfirmation...
task executed 0x00f1e1ebeff7df0cf7bd1b27b478085d842b7e0af99478a939b2a68ed6ba9c23
test gelato_api::tests::execute_transaction ... ok
```

Check the executed transaction on Görli: <https://goerli.etherscan.io/tx/0x00f1e1ebeff7df0cf7bd1b27b478085d842b7e0af99478a939b2a68ed6ba9c23>.

### CODE REVIEW

**Comment by vkgnosis** [comment]
Is it useful to add this when we currently can't use Gelato because require knowing the submitting account and nonce to associate auctions with their tx hashes?
There are some ideas on how to work around that but I don't think this is usable until then.

**Comment by nlordell** [comment]
> There are some ideas on how to work around that but I don't think this is usable until then.

We plan on experimenting with it on test-nets and layer 2s where solver rewards aren't needed.

**Comment by nlordell** [comment]
> For the submitter - I can add a `assert!(chain_id != 1)` check to prevent from accidental misuse.

Actually, this isn't needed because the `SolverTrampoline` contract isn't available on any other network at the moment (and needs to be an allow-listed solver anyway). Without this contract, relayed settlements don't work.

File: crates/shared/src/gelato_api.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+    pub fn with_url(factory: &HttpClientFactory, base: Url, api_key: String) -> Self {
+        Self {
+            client: factory.create(),
+            base,
+            api_key,
+        }
+    }
+
+    pub fn from_env() -> Result<Self> {
```
Comment (sunce86) [inline crates/shared/src/gelato_api.rs:41]: Can we add `#[cfg(test)]` here?

**Review by sunce86 — APPROVED** [review]
I've noticed there is no gas price in the request sent to API.
After reading the docs, saw that there is a gas price oracle we are not using, and also we don't define a cap on maximum cost we are willing to pay for a transaction.

Code LG overall. I think we can address this concerns in the follow up PR (if needed)

**Review by vkgnosis — APPROVED** [review]

File: crates/shared/src/gelato_api.rs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+/// A call to send to the Gelato relay network and to be executed onchain.
+#[serde_as]
+#[skip_serializing_none]
+#[derive(Clone, Debug, Default, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Call {
+    pub chain_id: u64,
+    pub target: H160,
+    #[serde(with = "model::bytes_hex")]
+    pub data: Vec<u8>,
```
Comment (MartinquaXD) [inline crates/shared/src/gelato_api.rs:92]: `Debug` should probably rather print the hex string presentation here. See PR [here](https://github.com/cowprotocol/services/pull/836/files) for the required changes.

**Review by MartinquaXD — APPROVED** [review]

File: crates/shared/src/gelato_api.rs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+            api_key,
+        }
+    }
+
+    pub fn from_env() -> Result<Self> {
+        Ok(Self::new(
+            &HttpClientFactory::default(),
+            env::var("GELATO_API_KEY")?,
+        ))
+    }
```
[the remark this answers is not in the document]
Comment (nlordell) [reply crates/shared/src/gelato_api.rs:46]: This is only used for testing. I will add a `#[cfg(test)]` to clarify this.

**Pull request merged**